#4766: Unable to generate RouteUrl for webapi

jtkech opened the pull request:
Fixes #4766

And maybe other related issues

> **In ShellRoute.cs, fixes the priority issue**

When using an UrlHelper extension, you needed to define your own HttpRouteDescriptor with a higher priority than other desciptors in the same area...

> **In StandardExtensionHttpRouteProvider.cs, defines default httproute names**

This by using the extension path that is equal to the extension name if no some special characters (e.g spaces if multiples words). Otherwise, it is equal to the extension Id (the extension folder name).

Then, for the http route name you can use "ModuleName.Api". If you can't, depending on the module name, then you can use "FolderName.Api"... This with an UrlHelper extension without having to specify the area name...

> **In RoutePublisher.cs, prevent the program to fail if duplicate route names**

There is a checking before the insertion in the final RouteCollection, but before there are 2 others internal RouteCollection where this is not done. Because the route name is useless for these 2 collections, we can pass an empty string

Best

jtkech commented:
@sebastienros 2 last points, maybe not useful

**For url generation** with url helpers `RouteUrl, HttpRouteUrl...`, when the controller / action are not specified, the current request values are used as defaults (e.g `.../Item/Display/x`). The possible Defaults values defined in the route descriptor are not used.

1. I've tried it, it's the same for all routes, not specific to http routes, and it seems to be inherent to the current behavior of the asp.net stack. I've found a simple fix that works, but I'm not sure what is the correct behavior. Maybe good to do this but only for routes that have a given name... Let me know.

2. In the default http descriptor per area, there is a default controller value `controller = "api"`. If 1. is not implemented, this is useless for url generation with url helpers. And for url mapping we would need to declare a controller with the not friendly name `apiController` (we can't use `ApiController`). So maybe another name, or using the first http controller name per area. I've tried this solution, it works but would be more useful if 1. is implemented. So, let me know, for now I've removed this default controller value.

Best

sebastienros commented on src/Orchard/WebApi/Routes/StandardExtensionHttpRouteProvider.cs:
```diff
@@ -22,9 +22,10 @@ public class StandardExtensionHttpRouteProvider : IHttpRouteProvider {
                 yield return new HttpRouteDescriptor {
+                    Name = displayPath + ".Api",
                     Priority = -10,
                     RouteTemplate = "api/" + displayPath + "/{controller}/{id}",
```
Isn't that a breaking change?

jtkech replied:
@sebastienros,

As I remember, only in the rare case where someone uses its own custom HttpRouteDescriptor, with a named route, with the same default name defined here `ModuleName.Api`, with a priority <= -10, and uses this route name in an urlhelper extension...

If so, you're right, he may have to change e.g the priority. But the program will not fail for a duplicate route name (fixed in this PR). And, because of the priority issue (fixed in this PR), most of the time when people define their own HttpRouteDescriptor, they use a priority > -10...

But, if you want, I can revert this change about default httproute names, and keep only the fixes about priority and duplicate route names... Let me know.

Best

Piedone commented:
Line endings in some files were fixed in the Orchard source, that's why this PR now has merge conflicts.

Please sync your fork with the Orchard repo and merge the target branch of this PR into your branch (for a guide see: https://help.github.com/articles/syncing-a-fork/ or alternatively if you use GitExtensions you can simply add https://github.com/OrchardCMS/Orchard.git under Repository/Remote repositories, do a fetch all and then merge the target branch into your branch).

If you have files appearing changed in your working directory that you didn't change but were actually touched by git to normalize line endings then you don't need to commit those (these were also fixed): clear the contents of the .gitattributes file in the repository root or remove the file, then do the merge. The changed files won't appear now (.gitattributes instructs git to care about line endings) but please don't commit the .gitattributes change (you can simply reset that change).

There shouldn't be too many merge conflicts and most of those should be resolved automatically by the merge tool (in GitExt hit Start merge tool and it will be able to do the merges).

Sorry about the inconvenience. On the upside this leftover from the hg to git conversion is now gone :-).

If you need further help just ping me here.

jtkech commented:
Sync done

sebastienros commented:
@dcinzona mentions that this PR doesn't fix the issue.

jtkech commented:
@sebastienros Ok, i think it was fixing an issue for me, but i don't remember the details. I will redo some tests on the last dev asap. @dcinzona can you describe again your exact issue? Then i could try to fix it. Thanks.

dcinzona commented:
I will simplify our issue:
We have 2 modules:
Module A adds an API route for an api controller `/api/{controller}/{action}` where controller is `FilesApiController`
Mobule B does the same thing where controller is `MembershipController`.
Both have a priority of -10 and a unique name and area.
Neither work.

jtkech commented:
Just re-tried step by step in the last dev and 1.10, and all work as described here and in the related issue #4766 when applying this PR. Then e.g you don't need to declare an `HttpRouteDescriptor` and you can use the default route name (not required) in place of the area parameter with an UrlHelper:

    <a href=@Url.HttpRouteUrl("ModuleName.Api", new { controller = "Test", p1 = "Test" })>Test1</a>

But this is based on the `StandardExtensionHttpRouteProvider.cs` `RouteTemplate` (here as before):

    RouteTemplate = "api/" + displayPath + "/{controller}/{id}",

So, i can understand the use case of @dcinzona because if you don't want to use this pattern, e.g without the `displayPath` and / or with an additional `{action}` segment, indeed you need to define your own `HttpRouteDescriptor`, and with a priority > -10 (or a specific route name, it also works).

But this is not related to the original priority issue with another regular descriptor in your module. This is because, in this case, you have to override the default `HttpRouteDescriptor` registered for us.

E.g, by completely removing the `StandardExtensionHttpRouteProvider` class, i can use a specific route pattern without route name and with a priority of -10 (and with the fix applied).

So, for me all work but maybe we also need to change the default http route pattern, e.g with {action} if it's the most commonly used. Here i think that the `displayPath` is useful to differentiate modules. Or completely remove this default registration and let people define their own http descriptors.

But maybe some people are already using the default registered http route template, e.g in a module that doesn't have another regular descriptor that create a priority issue.

Best.

jtkech commented:
@sebastienros here a summary of the 3 points related to this PR

> **In RoutePublisher.cs, prevent the program to fail if duplicate route names**

Just seen that 2 unit tests e.g `SameNameTwiceCausesExplosion()` expect an exception in case of duplicate names. Do i need to revert this change, or adapt the unit tests?

> **In StandardExtensionHttpRouteProvider.cs, defines default httproute names**

Just to provide a default route name based on the extension name. Not to fix an issue here but maybe useful. I can revert this change if you want.

> **In ShellRoute.cs, fixes the priority issue**

Fix the priority issue we have sometimes between an `http descriptor` and a `non http descriptor`. This when you use the default `RouteTemplate = "api/" + displayPath + "/{controller}/{id}"`.

But if you don't want to use this default pattern, e.g with an `{action}` segment, then again, you have to define your own `http descriptor` with a route name OR a priority > -10, but here it's to override the `default http descriptor` registered for us.

One possibility would be to change the default route template or maybe better completely remove this default registration. But may breaks those which are using this default without declaring their own http descriptors (maybe not so much).

@dcinzona, not sure, but if you have time, try to only apply the fix in `ShellRoute.cs`, remove the `StandardExtensionHttpRouteProvider` class definition, and tell me what happens.

Best

jtkech commented:
@sebastienros last summary of the 3 points related to this PR

- **Prevent from failling on duplicate route names**: Reverted because of 2 unit tests that expect an exception in this case. let me know if, instead, we have to adapt unit tests.

- **Default httproute names** : Just to provide a default route name based on the extension name. Not to fix an issue but maybe useful. I can also revert this change if you want.

- **Priority issue**: Fix the priority issue when you have a `non http descriptor` and you use  the `default http descriptor` whose route template is `"api/" + displayPath + "/{controller}/{id}"`.

If you don't want to use the default template, e.g with an `{action}` segment, then again, you have to define your own `http descriptor` with a route name OR a priority > -10, but here it's to override the `default http descriptor` registered for us. It's not the priority issue with a `non http descriptor`.

- **One possibility** would be to change the default route template with a more commonly used. Then, use it without declaring your own `http descriptor`, and without the priority issue because of the fix.

-  **Another possibility** is to completely remove the default registration. Then, always define you own `http descriptor`. But this would break people who use the default registration (maybe not so many).

Squashed.

Best

jtkech commented:
Because the recommended way to use web api is to use standard http verbs (Get, Post, Put, Delete...), in place of using an {action} segment, the default route template (currently used in the code), seems to be ok.

So, i removed the 2 last possibilities in my previous summary. See the UPDATE in my previous comment.

Best.
